framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java

```diff
@@ -768,6 +760,103 @@ public static AnnotatedTypeMirror leastUpperBound(
     return new AtmLubVisitor(atypeFactory).lub(type1, type2, lubTypeMirror);
   }
 
+  /**
+   * Returns the glb of two annotated types.
+   *
+   * @param atypeFactory the AnnotatedTypeFactory
+   * @param type1 annotated type
+   * @param type2 annotated type
+   * @return the glb of type1 and type2
+   */
+  public static AnnotatedTypeMirror greatestLowerBound(
+      AnnotatedTypeFactory atypeFactory, AnnotatedTypeMirror type1, AnnotatedTypeMirror type2) {
+    Types types = atypeFactory.types;
+    if (types.isSubtype(type1.getUnderlyingType(), type2.getUnderlyingType())) {
+      return glbSubtype(atypeFactory.getQualifierHierarchy(), type1, type2);
+    } else if (types.isSubtype(type2.getUnderlyingType(), type1.getUnderlyingType())) {
+      return glbSubtype(atypeFactory.getQualifierHierarchy(), type2, type1);
+    }
+
+    TypeMirror glbJava =
+        TypesUtils.greatestLowerBound(
+            type1.getUnderlyingType(), type2.getUnderlyingType(), atypeFactory.getProcessingEnv());
+
+    if (types.isSameType(type1.getUnderlyingType(), glbJava)) {
+      return glbSubtype(atypeFactory.getQualifierHierarchy(), type1, type2);
+    } else if (types.isSameType(type2.getUnderlyingType(), glbJava)) {
+      return glbSubtype(atypeFactory.getQualifierHierarchy(), type2, type1);
+    }
+
+    if (glbJava.getKind() != TypeKind.INTERSECTION) {
+      // If one type isn't a subtype of the other, then GLB must be an intersection.
+      throw new BugInCF(
+          "AnnotatedTypes#greatestLowerBound: unexpected java type: %s. type1: %s, type2: %s",
+          glbJava, type1, type2);
+    }
+    QualifierHierarchy qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+    Set<AnnotationMirror> setA =
+        AnnotatedTypes.findEffectiveLowerBoundAnnotations(qualifierHierarchy, type1);
+    Set<AnnotationMirror> setB =
+        AnnotatedTypes.findEffectiveLowerBoundAnnotations(qualifierHierarchy, type2);
+    Set<? extends AnnotationMirror> glbAnno = qualifierHierarchy.greatestLowerBounds(setA, setB);
+
+    AnnotatedIntersectionType glb =
+        (AnnotatedIntersectionType) AnnotatedTypeMirror.createType(glbJava, atypeFactory, false);
+
+    for (AnnotatedTypeMirror bound : glb.getBounds()) {
+      if (types.isSameType(bound.getUnderlyingType(), type1.getUnderlyingType())) {
+        atypeFactory.replaceAnnotations(type1, bound);
+      } else if (types.isSameType(bound.getUnderlyingType(), type2.getUnderlyingType())) {
+        atypeFactory.replaceAnnotations(type2, bound);
+      } else {
+        throw new BugInCF(
+            "Neither %s nor %s is one of the intersection bounds in %s. Bound: %s",
+            type1, type2, bound, glb);
+      }
+    }
+
+    glb.addAnnotations(glbAnno);
+    return glb;
+  }
+
+  /**
+   * Returns the annotated greatest lower bound of {@code subtype} and {@code supertype}, where the
```

Comment:
This method only manipulates the primary annotations.  The result might not be a subtype of the argument `supertype` (because of annotations other than primary annotations).  (This seems possible at the third and fourth call sites.)  Is this a bug or intended behavior?  Or is it impossible?

Reply:
This is the intended behavior.  I've renamed greatestLowerBound to annotatedGLB and added a comment explaining the algorithm.   However, there is a missing error related to this.  See [checker/tests/nullness/WildcardGLB.java](https://github.com/typetools/checker-framework/blob/ef727b2946a7e1332fba1934d0d940765dedc0ef/checker/tests/nullness/WildcardGLB.java).